tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -312,6 +333,160 @@ private int WriteSpanLink(ref byte[] bytes, int offset, in SpanModel spanModel)
+                            offset += MessagePackBinary.WriteInt32(ref bytes, offset, 4);
+                            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _arrayValueFieldBytes);
+                            offset += MessagePackBinary.WriteArrayHeader(ref bytes, offset, arrayVal.Length);
+
+                            foreach (var item in arrayVal)
+                            {
+                                offset += MessagePackBinary.WriteMapHeader(ref bytes, offset, 2);
+                                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _typeFieldBytes);
+                                offset += WriteEventAttribute(ref bytes, offset, item);
+                            }
+
+                            attrCount++;
+                        }
+                    }
+
+                    if (attrCount > 0)
+                    {
+                        MessagePackBinary.WriteMapHeaderForceMap32Block(ref bytes, attributeCountOffset, (uint)attrCount);
+                    }
+                    else
+                    {
+                        offset = attributeCountOffset - _attributesBytes.Length;
+                    }
+                }
+            }
+
+            return offset - originalOffset;
+        }
+
+        private int WriteEventAttribute(ref byte[] bytes, int offset, object value)
+        {
+            var originalOffset = offset;
+
+            if (value is string stringVal)
+            {
+                offset += MessagePackBinary.WriteInt32(ref bytes, offset, 0);
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _stringValueFieldBytes);
+                offset += MessagePackBinary.WriteString(ref bytes, offset, stringVal);
+            }
+            else if (value is bool boolVal)
+            {
+                offset += MessagePackBinary.WriteInt32(ref bytes, offset, 1);
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _boolValueFieldBytes);
+                offset += MessagePackBinary.WriteBoolean(ref bytes, offset, boolVal);
+            }
+            else if (value is int intValue)
+            {
+                offset += MessagePackBinary.WriteInt32(ref bytes, offset, 2);
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _intValueFieldBytes);
+                offset += MessagePackBinary.WriteInt64(ref bytes, offset, intValue);
+            }
+            else if (value is double doubleVal)
+            {
+                offset += MessagePackBinary.WriteInt32(ref bytes, offset, 3);
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _doubleValueFieldBytes);
+                offset += MessagePackBinary.WriteDouble(ref bytes, offset, doubleVal);
+            }
+
+            return offset - originalOffset;
+        }
+
+        private int WriteJsonEvents(ref byte[] bytes, int offset, in SpanModel spanModel)
```

Comment:
I would love to benchmark this particular method to make sure we're super performant, but I don't think that we need to block on it

Reply:
I could actually see you re-using the previous `ActivityEventConverter` class and just porting that logic over here

Reply:
Updated here:
https://github.com/DataDog/dd-trace-dotnet/pull/6769/commits/324c4372c1d35f5b23bc9221643fab7d368516ae